pkg/virt-handler/vm.go

```diff
@@ -377,6 +379,14 @@ func (d *VirtualMachineController) startDomainNotifyPipe(domainPipeStopChan chan
 		return err
 	}
 
+	if util.IsNonRootVMI(vmi) {
+		err := diskutils.DefaultOwnershipManager.SetFileOwnership(socketPath)
```

Comment:
This `DefaultOwnershipManager` was meant to be called within the `virt-launcher` container. It's doing a lookup on the default user `qemu` to find the UID/GID. This isn't guaranteed to be accurate when called from virt-handler.

If this happens to work within the virt-handler pod because the qemu user exists, then that's an accident. There's no need for the qemu user to exist in virt-handler and I doubt it does for people who build their kubevirt containers outside of our default build system.

Reply:
The handler image already contains the qemu user. @rmohr can you comment on why it was added? 

Afaik, we don't have any guarantees around images or guidelines on what should they contain(user, libvirt version, etc). 
I find it pretty straight forward to add the user to any build. The alternative would be to change only permission for others or switching to the context of the launcher. 

WDYT?

Reply:
> The handler image already contains the qemu user. @rmohr can you comment on why it was added?

That was mostly an efficiency decision. virt-handler uses the same base image as virt-launcher. It was just easy to do.

Reply:
I think the point I'm trying to make is that we can't depend on virt-handler having the qemu user. It does right now, but that's not necessarily on purpose.

If the UID/GID for qemu is stable (i think it is always 107) then is there a reason not to use the standard qemu UID/GID as a default instead of looking the user up?

Reply:
If I may add here, I am not actually sure that there are some 'standard' UID/GID for qemu user. It really depends and varies across distros. What about passing the UID/GID as cmdline arguments to `virt-handler` instead? They can be defaulted to 107 but still there will be a way to change those.

> There's no need for the qemu user to exist in virt-handler and I doubt it does for people who build their kubevirt containers outside of our default build system.

Hm... actually some time ago we faced an issue with our custom container images. There was an error from `virt-handler` about missing `qemu` user:
```json
{"component":"virt-handler","level":"error","msg":"Couldn't set Ownership on /var/run/kubevirt-private/vmi-disks/disk0/disk.img: failed to look up user qemu: user: unknown user qemu","pos":"host-disk.go:80","reason":"failed to look up user qemu: user: unknown user qemu","timestamp":"2021-03-19T06:24:40.295687Z"}
```
It was coming from `hostdisk.ReplacePVCByHostDisk`. This function is called in several places in `pkg/virt-handler/vm.go`, e.g.:

https://github.com/kubevirt/kubevirt/blob/1181ef70eb4f30053d4aeccdb24435951832c94e/pkg/virt-handler/vm.go#L2418

After that we explicitly added the qemu user to `virt-handler` container. Previously it wasnt there.

Reply:
> I was almost sure we need the qemu in handler.

there's some history here. We've tried  hard to not have virt-handler depend on libvirt/qemu being present. There's a separation of concerns here between handler/launcher that we've been trying to observe by keeping libvirt out of the picture for virt-handler. It's an accident that those dependencies made it into the virt-handler container. 


> @davidvossel
With that can we agree that it's fine for this PR to use it and do a follow-up that will maybe remove the dependency on the qemu in handler? We can do the runtime detection of non-root uid and actually add a configuration knob to change it.


long term, here's what I'd recommend considering

Virt-handler pod executes the virt-launcher container as an init container to dump some libvirt related stuff. Have that init container dump the qemu uid/gid into a file somewhere that virt-handler can read on initialization. For reference, this is the entry point script that the init container calls, https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-launcher/node-labeller/node-labeller.sh 

load the qemu uid/gid at virt-handler startup and fail early if the file's not found. 

> After that we explicitly added the qemu user to virt-handler container.

@vasiliy-ul so, is the user added regardless if the qemu/libvirt dependencies exist in the container?  If so, i'd be okay with doing the look up fo the qemu user in this PR (and following up wiht something better later) if it has nothing to do with these accidental dependencies existing.

Reply:
> there's some history here. We've tried hard to not have virt-handler depend on libvirt/qemu being present. There's a separation of concerns here between handler/launcher that we've been trying to observe by keeping libvirt out of the picture for virt-handler. It's an accident that those dependencies made it into the virt-handler container.

I think that makes sense. I also got impression that it was rather a temporary solution with qemu user in virt-handler.

> long term, here's what I'd recommend considering
>
>Virt-handler pod executes the virt-launcher container as an init container to dump some libvirt related stuff. Have that init container dump the qemu uid/gid into a file somewhere that virt-handler can read on initialization. For reference, this is the entry point script that the init container calls, https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-launcher/node-labeller/node-labeller.sh
>
>load the qemu uid/gid at virt-handler startup and fail early if the file's not found.

 :+1: 

> @vasiliy-ul so, is the user added regardless if the qemu/libvirt dependencies exist in the container? If so, i'd be okay with doing the look up fo the qemu user in this PR (and following up wiht something better later) if it has nothing to do with these accidental dependencies existing.

Yes, we add only the user. Currently we even set UID/GID to be `107` to avoid possible issues. I think the look up should work fine for now. Though yes, its better to address it in the follow-up (e.g. like proposed with the init virt-launcher container).